pkg/services/oauthtoken/oauth_token.go

```diff
@@ -107,6 +114,89 @@ func (o *Service) IsOAuthPassThruEnabled(ds *datasources.DataSource) bool {
 	return ds.JsonData != nil && ds.JsonData.Get("oauthPassThru").MustBool()
 }
 
+// HasOAuthEntry returns true and the UserAuth object when OAuth info exists for the specified User
+func (o *Service) HasOAuthEntry(ctx context.Context, usr *user.SignedInUser) (bool, *models.UserAuth) {
+	if usr == nil {
+		// No user, therefore no token
+		return false, nil
+	}
+
+	authInfoQuery := &models.GetAuthInfoQuery{UserId: usr.UserID}
+	err := o.AuthInfoService.GetAuthInfo(ctx, authInfoQuery)
+	if err != nil {
+		if !errors.Is(err, user.ErrUserNotFound) {
+			logger.Error("failed to get OAuth token for user", "userId", usr.UserID, "username", usr.Login, "error", err)
+		}
+		return false, nil
+	}
+	if !strings.Contains(authInfoQuery.Result.AuthModule, "oauth") {
+		return false, nil
+	}
+	return true, authInfoQuery.Result
+}
+
+// TryTokenRefresh returns an error in case the OAuth token refresh was unsuccessful
+// It uses a singleflight.Group to prevent getting the Refresh Token multiple times for a given User
+func (o *Service) TryTokenRefresh(ctx context.Context, usr *models.UserAuth) error {
+	lockKey := fmt.Sprintf("oauth-refresh-token-%d", usr.UserId)
+	logger.Debug("starting a singleflight for getting a new access token", "key", lockKey)
+	_, err, _ := o.sf.Do(lockKey, func() (interface{}, error) {
```

Comment:
nit:

I'm not a huge fan of abbreviated variables like `o` or `sf` since they degrade readability and we can lose context. I won't block a PR because of this. However, let's keep in mind this is a public repo.

Reply:
Changed `sf` but for the receiver I left the `o`, because it's common to use only a word for the receiver's name.